### PR TITLE
[SPARK-21256] [SQL] Add withSQLConf to Catalyst Test

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?
SQLConf is moved to Catalyst. We are adding more and more test cases for verifying the conf-specific behaviors. It is nice to add a helper function to simplify the test cases.

### How was this patch tested?
N/A

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/CatalystTestUtils.scala [hunk 1]
```diff
@@ -37,4 +40,32 @@ object DateTimeTestUtils {
       DateTimeUtils.resetThreadLocals()
     }
   }
+
+  /**
+   * Sets all SQL configurations specified in `pairs`, calls `f`, and then restore all SQL
+   * configurations.
+   */
+  protected def withSQLConf(pairs: (String, String)*)(f: => Unit): Unit = {
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/CatalystTestUtils.scala]: This is basically from SQLTestUtils

Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/CatalystTestUtils.scala]: Hi, @gatorsmile .
Then, what about removing `withSQLConf` from `SQLTestUtils` instead of having duplicated logic?
Since this is `trait`, I think we can do that in this PR, too.

**Comment by dongjoon-hyun** [comment]
+1, LGTM (with one minor optional comment).

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/CatalystTestUtils.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/CatalystTestUtils.scala]: Sure.

**Comment by rxin** [comment]
Can we minimize the change by just adding this method to PlanTest? It's not that many lines of code.

**Comment by gatorsmile** [comment]
Sure

File: sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala [hunk 2]
```diff
@@ -89,28 +92,11 @@ private[sql] trait SQLTestUtils
...
-    try f finally {
-      keys.zip(currentValues).foreach {
-        case (key, Some(value)) => spark.conf.set(key, value)
-        case (key, None) => spark.conf.unset(key)
-      }
-    }
+  protected override def withSQLConf(pairs: (String, String)*)(f: => Unit): Unit = {
+    // ensure spark's session has been initialized and set to the current SQLConf.confGetter
+    // TODO: fix the multi-session supports for SQLConf.confGetter
+    SQLConf.setSQLConfGetter(() => spark.sessionState.conf)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala]: Ideally, directly calling `withSQLConf` of `PlanTest` should work. That means, this line is not needed. However, it does not work in the current `SQLConf.confGetter`. Another PR is needed to fix that issue.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala [hunk 3]
```diff
@@ -28,8 +29,9 @@ import org.apache.spark.sql.internal.SQLConf
 /**
  * Provides helper methods for comparing plans.
  */
-abstract class PlanTest extends SparkFunSuite with PredicateHelper {
+trait PlanTest extends SparkFunSuite with PredicateHelper {
 
+  // TODO(gatorsmile): remove this from PlanTest and all the analyzer/optimizer rules
   protected val conf = new SQLConf().copy(SQLConf.CASE_SENSITIVE -> true)
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: This line should not be needed. We can use the global SQLConf for Catalyst package.

**Comment by gatorsmile** [comment]
cc @cloud-fan @rxin

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: what stops us from doing so now?

Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: That requires changing the source codes of `analyzer/optimizer`. We have a bug in multi-session supports. We need to fix that before changing the source codes of `analyzer/optimizer`

Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: It sounds like this is not a bug. We have to call `setActiveSession` to ensure the current session is pointing to the right conf.

Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: Let me submit a separate PR for removing SQLConf from Analyzer and Optimizer.

Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: The PR is pretty big. It requires `withSQLConf` : )

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM, pending jenkins

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merging to master.
